### PR TITLE
FW-234 SCons build for Atmel SAMR21-XPlained Pro

### PR DESCRIPTION
**Pull request by mikehaben**
Amended SCons scripts, and added driver-source, header and linker-script files to enable SCons build for the Atmel  SAMR21-Xplained Pro board. 
Example Windows command-line to build the OpenWSN project:
C:\Python27\scons.bat board=samr21_xpro toolchain=armgcc oos_openwsn

### CODE REVIEW

**Comment by changtengfei** [comment]
@mikehaben , I added a travis job for SAMR21 so travis will build your code. Hope you don't mind. I have tested your code on my side, and it works.  Thanks a lot for the contribution!

**Comment by mikehaben** [comment]
Great, glad to be able to contribute something to this project.  And get to understand Github a bit better...

**Comment by changtengfei** [comment]
@mikehaben , the travis build doesn't pass, this is caused the different platform problem:
```
bsp/boards/samr21_xpro/drivers/inc/samr21_sysctrl.h:247:2: error: 'for' loop initial declarations are only allowed in C99 or C11 mode
  for (uint32_t i = 0; i < EIC_INST_NUM; i++) {
```

The uint32_t i should put at the beginning of the function. Could you check all the code which are added into the repository to fix the similar issue?

File: SConscript [hunk 1]
```diff
@@ -302,6 +302,8 @@ elif env['toolchain']=='armgcc':
         env.Append(CCFLAGS       = '-Ibsp/boards/samr21_xpro/cmsis/inc')
         env.Append(CCFLAGS       = '-D__SAMR21G18A__')
         env.Append(CCFLAGS       = '-Ibsp/boards/samr21_xpro/SAMR21_DFP/1.0.34/include')
+        env.Append(CCFLAGS       = '-std=c99')
+        -std=c99
```
Comment (changtengfei) [inline SConscript]: Unexpected line code?

**Comment by changtengfei** [comment]
@mikehaben , congrats to make it work! I just merged this PR. 
Now I am trying to close the related issue at: https://openwsn.atlassian.net/browse/FW-234
But there are two more PRs associated it, could you take a look? If they are duplicated, you can close them. 

Thanks very much for the work!

**Comment by mikehaben** [comment]
Should be all closed now.  A good end to the week!

**Comment by iseitani** [comment]
@mikehaben The produced file is a .bin file, right?

**Comment by iseitani** [comment]
Dear @mikehaben  & @changtengfei , please correct me if I am wrong but the output of the file <elf_file> in build/samr21_xpro_armgcc/projects/common prints a 64bit elf file, the samr21 has a 32bit processor, by just replacing the samr21_xpro to openmote-cc2538 on the scons build command produces a 32bit elf file

**Comment by changtengfei** [comment]
@iseitani , I am afraid I don't understand. How it generates a 64bit version file and what you mean by replacing the samr21_xpro to openmote-cc2538 on scons build? @mikehaben , you probably know more thing than me. Do you understand?

**Comment by iseitani** [comment]
@changtengfei   The command to compile openwsn for samr21 is :board=samr21_xpro toolchain=armgcc oos_openwsn[1] (as mentioned at the beginning of the thread) the output of this command are .elf, .ihex files located in the directory : openwsn-fw/build/samr21_xpro_armgcc/projects/common[2]. In the command "board=samr21_xpro toolchain=armgcc oos_openwsn" if I change the board (i.e. instead of samr21_xpro to openmote-cc2538[board=openmote-cc2538 toolchain=armgcc oos_openwsn]) in [2] I get a 32 bit elf file while if I execute [1] in the correspondng directory I will get a 64bit elf file

**Comment by changtengfei** [comment]
@iseitani so you mean it should be 32bit elf file rather than 64bit elf file for image generated through scons command for samr21. @mikehaben any comment on this?

**Comment by mikehaben** [comment]
Sorry, no idea why it would decide to generate 64-bit executables for any of the supported processors.

**Comment by changtengfei** [comment]
@mikehaben does the 64bit file allow you to download to the micro-controller and work well also. If so, I think it should be fine, right?

**Comment by iseitani** [comment]
@changtengfei  The .ihex file produced by [1] can be uploaded in the device by using the Atmel Studio, but  it seems not to work, as after flashing the device and just powering it, it's not printed as a neighbour in OV, I don't know if the problem is that I am using as a sink an OpenBase device.

[1] board=samr21_xpro toolchain=armgcc oos_openwsn

**Comment by mikehaben** [comment]
If you're using Atmel Studio, why not build the A-Studio project? (03oos_openwsn.atsln, .cproj)

**Comment by changtengfei** [comment]
@iseitani , what the following means?

> it's not printed as a neighbour in OV.

**Comment by iseitani** [comment]
@changtengfei I start the openvisualizer with the scons runweb command, on my pc I have attached an openmote-cc2538 Openbase device which serves as a sink. On the webUI of the OpenVisualizer under the  tab neighbours I am not able to see the ID of the samr21 device, while if I switch on an openmote-cc2538 OpenBattery device I will immediately see the ID of the OpenBattery device in that list. Also, in the terminal, where the openvisualizer is running, I don't receive any messages from the samr21 device.I initially thought that the reason why I was not able to establish a connection between the OpenBase and the samr21  was the 64bit produced executable file.

**Comment by changtengfei** [comment]
@iseitani, I see your point. Do you have Atmel studio installed and whether it works for you? If that works, it would be the problem of scons build.

@mikehaben, built from scons is a tool easier for all user developing, no matter they have Atmel studio installed or not. All other platform supports scons currently. Does it work on you side currently? It's fine if it doesn't support right now. It would be good to have this support in the future.

**Comment by mikehaben** [comment]
I haven't worked on this for a few months, so don't have equipment set up to test with.
I think that the toolchain generating a 64-bit ELF is wrong, the SAMR21 doesn't execute 64-bit code.  If he has Atmel Studio installed, I would check that the code it generates runs OK, and then make the SCons build use the same GCC toolchain as A. Studio.

**Comment by changtengfei** [comment]
Since the code has been merged. I assume iseitani is using the same code you have developed? Could you confirm @iseitani ? 
Also I just call the scons to build the samr21 board code and get a 32bit version exe file. Is that possible some tools you are using is 64bit version? Like scons or python?

**Comment by iseitani** [comment]
Yes, I am using the latest master rep. And the toolchain is the one downloaded by the script on the "get started linux". So the configuration on both of us is the same. Based on[1] I suppose that you used Windows, right? I attach the output of both the ls -la command in the  "openwsn-fw/build/samr21_xpro_armgcc/projects/common" directory[2] and the  file comand on the .elf file[3].
However, the point is that even when I compiled and flashed openwsn with Atmel Studio I was still unable to see as a neighbour in the web ui of openvisualizer the samr21 board.

[1] https://openwsn.atlassian.net/wiki/questions/42729493/binary-generated-with-scons
[2] https://goo.gl/MsNBlL
[3] https://goo.gl/KmY4C4

**Comment by changtengfei** [comment]
@iseitani yes, I am using windows. So you are using a 64bit version Linux? If you open a terminal (minicom e.g. baudrate= 115200), are you able to see something like: ~S......~ (dot is used for replacing any characters.)
@mikehaben I remember there was a problem with samr21 before for receiving serial data from openvisualizer, do you still remember? Any modification required on openvisualizer side?

**Comment by mikehaben** [comment]
No, no OpenVisualizer modification for serial-data to the SAMR21 (or SAMD21 that most of my work is on).

**Comment by changtengfei** [comment]
@mikehaben  I see. Thanks for clarification!

**Pull request merged**